Duckling/Types/Document.hs

```diff
@@ -125,9 +125,59 @@ data CharClass
 -- the reasonability of the match to actually be a word.
 isRangeValid :: Lang -> Document -> Int -> Int -> Bool
 isRangeValid = \case
+  AR -> arIsRangeValid
   ZH -> zhIsRangeValid
   _ -> defaultIsRangeValid
   where
+    arIsRangeValid :: Document -> Int -> Int -> Bool
+    arIsRangeValid doc start end =
+      ((start == 0 ||
+            isDifferent (doc ! (start - 1)) (doc ! (start))) &&
+        (end == length doc ||
+            isDifferent (doc ! (end - 1)) (doc ! (end)))) ||
+      -- Is Arabic proclitic?
+      (start == end - 1 &&
+        isArabicProclitic (doc ! start) &&
+        (start == 0 || isDifferent (doc ! (start - 1)) (doc ! start))) ||
+      (start == end - 2 &&
+        isArabicProclitic2 (doc ! start) (doc ! (start + 1)) &&
+        (start == 0 || isDifferent (doc ! (start - 1)) (doc ! start))) ||
+      -- Is preceeded by proclitic
+      (start /= 0 && isArabicProclitic (doc ! (start - 1)) &&
+        (end == length doc ||
+          isDifferent (doc ! (end - 1)) (doc ! (end)))) ||
+      -- Is Arabic enclitic?
+      (start == (end - 2) && isArabicEnclitic (doc ! start) (doc ! (end - 1)) &&
+        (end == length doc || isDifferent (doc ! end) (doc ! (end + 1)))) ||
+      -- Is followed by enclitic
+      ((start ==0 || isDifferent (doc ! (start - 1)) (doc ! (start))) &&
+        (end <= (length doc - 2) &&
+          isArabicEnclitic (doc ! (end)) (doc ! (end + 1))))
+      where
+        -- This list isn't exhasutive since Arabic have some diacritics and rarely used characters in Unicode
+        isArabic :: Char -> Bool
+        isArabic c = elem c ['ا', 'ب', 'ت', 'ة', 'ث', 'ج', 'ح', 'خ', 'د', 'ذ', 'ر', 'ز', 'س', 'ش', 'ص', 'ض', 'ط', 'ظ', 'ع', 'غ', 'ف', 'ق', 'ك', 'ل', 'م', 'ن', 'ه', 'ي', 'ء', 'آ', 'أ', 'إ', 'ؤ', 'ئ', 'ى']
```

Comment:
Or...

```suggestion
        isArabic c = elem c ['\1536' .. '\1791']
```

Works as follows (pardon Github RTL rendering):

```haskell
Prelude> elem 'و' ['\1536' .. '\1791']
True
Prelude> elem 'ؤ' ['\1536' .. '\1791']
True
Prelude> elem 'a' ['\1536' .. '\1791']
False
```

Reply:
This looks much better but actually I am not a fan of using this list since it contains lots of irrelevant characters that might cause problems.
If such list is used then you would consider characters like numerals and Arabic punctuation marks as digits which would break the rules and cause problems.
I prefer to have a restricted list of characters that can be augmented later to fix any bugs that might arise.
What do you think?
The long list of characters in the range:
```
؀
؁
؂
؃
؄
؅
؆
؇
؈
؉
؊
؋
،
؍
؎
؏
ؐ
ؑ
ؒ
ؓ
ؔ
ؕ
ؖ
ؗ
ؘ
ؙ
ؚ
؛
؜
؝
؞
؟
ؠ
ء
آ
أ
ؤ
إ
ئ
ا
ب
ة
ت
ث
ج
ح
خ
د
ذ
ر
ز
س
ش
ص
ض
ط
ظ
ع
غ
ػ
ؼ
ؽ
ؾ
ؿ
ـ
ف
ق
ك
ل
م
ن
ه
و
ى
ي
ً
ٌ
ٍ
َ
ُ
ِ
ّ
ْ
ٓ
ٔ
ٕ
ٖ
ٗ
٘
ٙ
ٚ
ٛ
ٜ
ٝ
ٞ
ٟ
٠
١
٢
٣
٤
٥
٦
٧
٨
٩
٪
٫
٬
٭
ٮ
ٯ
ٰ
ٱ
ٲ
ٳ
ٴ
ٵ
ٶ
ٷ
ٸ
ٹ
ٺ
ٻ
ټ
ٽ
پ
ٿ
ڀ
ځ
ڂ
ڃ
ڄ
څ
چ
ڇ
ڈ
ډ
ڊ
ڋ
ڌ
ڍ
ڎ
ڏ
ڐ
ڑ
ڒ
ړ
ڔ
ڕ
ږ
ڗ
ژ
ڙ
ښ
ڛ
ڜ
ڝ
ڞ
ڟ
ڠ
ڡ
ڢ
ڣ
ڤ
ڥ
ڦ
ڧ
ڨ
ک
ڪ
ګ
ڬ
ڭ
ڮ
گ
ڰ
ڱ
ڲ
ڳ
ڴ
ڵ
ڶ
ڷ
ڸ
ڹ
ں
ڻ
ڼ
ڽ
ھ
ڿ
ۀ
ہ
ۂ
ۃ
ۄ
ۅ
ۆ
ۇ
ۈ
ۉ
ۊ
ۋ
ی
ۍ
ێ
ۏ
ې
ۑ
ے
ۓ
۔
ە
ۖ
ۗ
ۘ
ۙ
ۚ
ۛ
ۜ
۝
۞
۟
۠
ۡ
ۢ
ۣ
ۤ
ۥ
ۦ
ۧ
ۨ
۩
۪
۫
۬
ۭ
ۮ
ۯ
۰
۱
۲
۳
۴
۵
۶
۷
۸
۹
ۺ
ۻ
ۼ
۽
۾
ۿ
```

Reply:
Makes perfect sense!